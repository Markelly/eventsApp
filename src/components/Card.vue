<template>
  <b-card no-body class="overflow-hidden">
    <b-row no-gutters class="card__row">
      <b-col md="4">
        <b-img :src="event.image" fluid-grow rounded />
      </b-col>
      <b-col md="8" class="card__content">
        <Date :date="event.date" />
        <div class="card__name" v-text="event.name"></div>
        <div class="card__price" v-text="event.price"></div>
        <div class="card__icons">
          <img :src="getRedirectIcon" @click="redirectToEventPage" data-icon-redirect />
          <img :src="getIcon" @click="saveFavorite" />
        </div>
      </b-col>
    </b-row>
  </b-card>
</template>

<script>
import { mapActions } from "vuex";
import { getImageContext } from "@/utils/stringUtils";
import Date from "@/components/Date";

export default {
  name: "Card",
  components: { Date },
  props: {
    event: {
      type: Object,
      required: true
    }
  },
  computed: {
    getIcon() {
      const icon = this.event.favorite ? "heart-solid" : "heart-regular";
      return getImageContext(icon);
    },
    getRedirectIcon() {
      return getImageContext("external-link");
    }
  },
  methods: {
    ...mapActions({
      addToFavorites: "addToFavorites"
    }),
    saveFavorite() {
      this.addToFavorites(this.event);
    },
    redirectToEventPage() {
      window.open(this.event.url);
    }
  }
};
</script>

<style scoped lang="scss">
@import "@/styles/events-app";

.card {
  background: #f4f7f6;
  border: solid 2px white;
  margin: 8px 0;

  &:hover {
    background: $color-white;
    border: solid 2px $color-light-green;
    color: $color-secondary;
  }

  &__row {
    align-items: center;
    width: 100%;
  }

  &__content {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    padding: 16px !important;

    @media (min-width: $breakpoint-tablet) {
      padding-left: 16px !important;
      text-align: left;
    }
  }

  &__name {
    font-size: 20px;
    padding: 4px;

    @media (min-width: $breakpoint-tablet) {
      font-size: 24px;
      padding: 8px 0;
    }
  }

  &__price {
    font-style: italic;
    padding: 4px;
    color: $color-dark-gray;
    font-size: 12px;

    @media (min-width: $breakpoint-tablet) {
      font-size: 16px;
    }
  }

  &__icons {
    margin-bottom: 8px;
    display: flex;
    justify-content: flex-end;

    & > * {
      cursor: pointer;
      padding-left: 24px;
    }

    @media (min-width: $breakpoint-tablet) {
      margin: 0 16px;
    }
  }
}
</style>
