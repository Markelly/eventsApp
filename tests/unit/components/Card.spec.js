import Card from "../../../src/components/Card";
import { createLocalVue, shallowMount } from "@vue/test-utils";
import BootstrapVue from "bootstrap-vue";

const localVue = createLocalVue();
localVue.use(BootstrapVue)

describe("<Card/>", () => {
  let wrapper;
  let event = { id: "1", favorite: false, date: { start: "2020-03-20" } };

  beforeEach(() => {
    wrapper = shallowMount(Card, {
      localVue,
      propsData: {
        event
      },
      computed: {
        getIcon: () => {
          return "foo";
        },
        getRedirectIcon: () => {
          return "bar";
        }
      }
    });
  });

  test("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should redirect user to event page", () => {
    window.open = jest.fn();
    wrapper.find("[data-icon-redirect]").trigger("click");
    expect(window.open).toHaveBeenCalled();
  });
});
