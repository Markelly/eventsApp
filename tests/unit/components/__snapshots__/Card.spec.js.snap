// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> should render the component 1`] = `
<b-card-stub bodytag="div" titletag="h4" subtitletag="h6" subtitletextvariant="muted" headertag="div" footertag="div" tag="div" nobody="true" class="overflow-hidden">
  <b-row-stub tag="div" nogutters="true" class="card__row">
    <b-col-stub md="4" tag="div">
      <b-img-stub fluidgrow="true" rounded="true" blankcolor="transparent"></b-img-stub>
    </b-col-stub>
    <b-col-stub md="8" tag="div" class="card__content">
      <date-stub date="[object Object]"></date-stub>
      <div class="card__name"></div>
      <div class="card__price"></div>
      <div class="card__icons"><img src="bar" data-icon-redirect=""> <img src="foo"></div>
    </b-col-stub>
  </b-row-stub>
</b-card-stub>
`;
